[GLib] WebKitWebHitTestResult should use composition instead of subclassing

aperezdc opened the pull request:
#### 71df87cb0f482d857f21d1d93d13d097b14eaa42
<pre>
[GLib] WebKitWebHitTestResult should use composition instead of subclassing
<a href="https://bugs.webkit.org/show_bug.cgi?id=251075">https://bugs.webkit.org/show_bug.cgi?id=251075</a>

Reviewed by Michael Catanzaro.

Make WebKitWebHitTestResult a final type that inherits directly from
GObject with the new 2022 GLib API, which uses a WebKitHitTestResult
instance internally. The methods which were previously inherited from
the base class are added, with their implementations forwarding to the
corresponding WebKitHitTestResult ones. While at it, WebKitHitTestResult
can now be made final and the _WebKitWebHitTestResultClass struct can be
removed from the public header.

* Source/WebKit/Shared/API/glib/WebKitHitTestResult.cpp: Make the type
  final with in new 2022 API.
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.cpp:
(webkitWebHitTestResultCreate): Arrange to instantiate the contained
WebKitHitTestResult with the new API.
(webkit_web_hit_test_result_get_context): Added, forwards its
implementation to the corresponding WebKitHitTestResult method.
(webkit_web_hit_test_result_context_is_link): Ditto.
(webkit_web_hit_test_result_context_is_image): Ditto.
(webkit_web_hit_test_result_context_is_media): Ditto.
(webkit_web_hit_test_result_context_is_editable): Ditto.
(webkit_web_hit_test_result_context_is_selection): Ditto.
(webkit_web_hit_test_result_get_link_uri): Ditto.
(webkit_web_hit_test_result_get_link_title): Ditto.
(webkit_web_hit_test_result_get_link_label): Ditto.
(webkit_web_hit_test_result_get_image_uri): Ditto.
(webkit_web_hit_test_result_get_media_uri): Ditto.
(webkit_web_hit_test_result_context_is_scrollbar): Ditto.
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.h.in:
  Add methods for the new 2022 API which cover the functionality which
  was previously inherited from WebKitHitTestResult.
* Source/WebKit/gtk/migrating-to-webkitgtk-6.0.md: Added migration note.

Canonical link: <a href="https://commits.webkit.org/259651@main">https://commits.webkit.org/259651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0093df5437ef741accdf46ec2a114beb757f6adb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105556 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/14645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/38452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/114807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16065 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/5868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111311 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/38452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/97850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/94080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/38452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7931 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/38452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/8318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/5868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14055 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/38452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6662 "Failed to push commit to Webkit repository") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/9951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos reviewed (commented):
I don't see the problem of using inheritance and leave the object derivable, but I'm not opposed to change it either. Instead of duplicating the API, why not just adding a getter for the internal hit test result instead? I don't see the point of not inheriting, but then pretend it's a derived class with same api.

aperezdc committed:
[GLib] WebKitWebHitTestResult should use composition instead of subclassing
https://bugs.webkit.org/show_bug.cgi?id=251075

Reviewed by Michael Catanzaro.

Make WebKitWebHitTestResult a final type that inherits directly from
GObject with the new 2022 GLib API, which uses a WebKitHitTestResult
instance internally. The methods which were previously inherited from
the base class are added, with their implementations forwarding to the
corresponding WebKitHitTestResult ones. While at it, WebKitHitTestResult
can now be made final and the _WebKitWebHitTestResultClass struct can be
removed from the public header.

* Source/WebKit/Shared/API/glib/WebKitHitTestResult.cpp: Make the type
  final with in new 2022 API.
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.cpp:
(webkitWebHitTestResultCreate): Arrange to instantiate the contained
WebKitHitTestResult with the new API.
(webkit_web_hit_test_result_get_context): Added, forwards its
implementation to the corresponding WebKitHitTestResult method.
(webkit_web_hit_test_result_context_is_link): Ditto.
(webkit_web_hit_test_result_context_is_image): Ditto.
(webkit_web_hit_test_result_context_is_media): Ditto.
(webkit_web_hit_test_result_context_is_editable): Ditto.
(webkit_web_hit_test_result_context_is_selection): Ditto.
(webkit_web_hit_test_result_get_link_uri): Ditto.
(webkit_web_hit_test_result_get_link_title): Ditto.
(webkit_web_hit_test_result_get_link_label): Ditto.
(webkit_web_hit_test_result_get_image_uri): Ditto.
(webkit_web_hit_test_result_get_media_uri): Ditto.
(webkit_web_hit_test_result_context_is_scrollbar): Ditto.
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.h.in:
  Add methods for the new 2022 API which cover the functionality which
  was previously inherited from WebKitHitTestResult.
* Source/WebKit/gtk/migrating-to-webkitgtk-6.0.md: Added migration note.

Canonical link: https://commits.webkit.org/259651@main
